internal/app/singularity/verify_ocsp.go

```diff
@@ -0,0 +1,237 @@
+)
+
+/*
+Online Certificate Status Protocol - OCSP
+
+OCSP responder is used to provide real-time verification of the revocation status of an X.509 certificate.
+RFC: https://www.rfc-editor.org/rfc/rfc6960
+*/
+
+const (
+	PKIXOCSPNoCheck = "1.3.6.1.5.5.7.48.1.5"
+)
+
+var errOCSP = errors.New("OCSP verification has failed")
+
+func OCSPVerify(chain ...*x509.Certificate) error {
+	// use the pool as an index for certificate issuers.
+	// fixme: we can drop this lookup if we assume that certificate N is always signed by certificate N+1.
+	pool := map[string]*x509.Certificate{}
+
+	for _, cert := range chain {
+		pool[string(cert.SubjectKeyId)] = cert
+	}
+
+	// recursively validate the certificate chain
+	for _, cert := range chain {
+		if err := validateCertificate(cert, pool); err != nil {
+			sylog.Warningf("OCSP verification has failed. Err: %s", err)
+			return errOCSP
+		}
+	}
+
+	return nil
+}
+
+func validateCertificate(cert *x509.Certificate, pool map[string]*x509.Certificate) error {
+	if len(cert.AuthorityKeyId) == 0 || string(cert.SubjectKeyId) == string(cert.AuthorityKeyId) {
+		sylog.Infof("skip self-signed certificate (%s)", cert.Subject.String())
+
+		return nil
+	}
+
+	/*---------------------------------------------------
+	 * Retrieve the CA who issued the certificate in question.
+	 *---------------------------------------------------*/
+
+	// firstly, look for the issuer in the pool of certificates.
+	issuer, exists := pool[string(cert.AuthorityKeyId)]
+	if !exists {
+		// if not found anywhere locally, try to download it
+		missingCerts, err := downloadCertsFromURLs(cert.IssuingCertificateURL...)
+		if err != nil {
+			return errors.Wrapf(err, "download cert error")
+		}
+
+		// if that does not work either, just abort
+		issuer, exists = missingCerts[string(cert.AuthorityKeyId)]
+		if !exists {
+			return errors.Errorf("cannot find issuer '%s'", cert.Issuer)
+		}
+	}
+
+	sylog.Infof("Validate: cert:%s  issuer:%s", cert.Subject.CommonName, issuer.Subject.CommonName)
+
+	/*---------------------------------------------------
+	 * Ask OCSP for the validity of signer's certificate.
+	 * Also make sure that the OCSP is trustworthy.
+	 *---------------------------------------------------*/
+	ocspCertificate, err := queryOCSP(cert, issuer)
+	if err != nil {
+		return errors.Wrapf(err, "OCSP Query")
+	}
+
+	if ocspCertificate != nil {
+		// The CA requires us to explicitly trust this certificate
+		// RFC-6960 Section: 4.2.2.2.1
+		for _, extension := range cert.Extensions {
+			if extension.Id.String() == PKIXOCSPNoCheck {
+				goto skipOCSPVerification
+			}
+		}
+
+		// avoid loops cause by misconfigured OCSP responders
+		if string(cert.SubjectKeyId) == string(ocspCertificate.SubjectKeyId) {
+			return nil
+		}
+
+		// make sure that the OCSP server is trustworthy
+		if err := validateCertificate(ocspCertificate, pool); err != nil {
+			return errors.Wrapf(err, "cannot verify OCSP server's certificate")
+		}
```

Comment:
I believe this only validates that the cert isn't revoked? I believe we need to call [Certificate.Verify](https://pkg.go.dev/crypto/x509#Certificate.Verify) as well to ensure the certificate is valid?

Reply:
We could add it, but I don't think that is mentioned in the [RFC](https://www.ietf.org/rfc/rfc2560.txt). 

Plus, it will be recursively validated by the OCSP server.

Reply:
> We could add it, but I don't think that is mentioned in the [RFC](https://www.ietf.org/rfc/rfc2560.txt).
> 
> Plus, it will be recursively validated by the OCSP server.

I believe `§2.2 Response` does mandate this?

> All definitive response messages SHALL be digitally signed. The key used to sign the response MUST belong to one of the following:
> 
>    -- the CA who issued the certificate in question
>    -- a Trusted Responder whose public key is trusted by the requester
>    -- a CA Designated Responder (Authorized Responder) who holds a specially marked certificate issued directly by the CA, indicating that the responder may issue OCSP responses for that CA

In the last case, [ocsp.ParseResponseForCert](https://pkg.go.dev/golang.org/x/crypto/ocsp#ParseResponseForCert) calls [Response.CheckSignatureFrom](https://pkg.go.dev/golang.org/x/crypto/ocsp#Response.CheckSignatureFrom), which validates that the supplied issuer signed the certificate, but does not validate that `resp.Certificate` is authorized to issue OCSP responses as indicated in §4.2.2.2. In other words, if `resp.Certificate` isn't `nil`, I believe we need to validate that it includes an an `ExtendedKeyUsage` of `id-ad-ocspSigning`?

Not sure if that's clear... this stuff is complicated!

Reply:
Aham. I see two parallel threads here.
- How do we know that the OCSP response has not been tampered with?
- How do we know that the OCSP responder is trustworthy? 

== Response Integrity ==
As you said, [ocsp.ParseResponseForCert](https://pkg.go.dev/golang.org/x/crypto/ocsp#ParseResponseForCert) calls [Response.CheckSignatureFrom](https://pkg.go.dev/golang.org/x/crypto/ocsp#Response.CheckSignatureFrom), which validates that the supplied issuer signed the certificate. 

If the returned certificate is not `nil`, the `verify_ocsp.go` goes through the logic for `Authorized Responders`

https://github.com/sylabs/singularity/blob/49bb33b0f81c84517d679e18e8324581fa93dd7d/internal/app/singularity/verify_ocsp.go#L91-L92

== Authorized Responders ==
The logic of `§4.2.2.2.` is already implemented and documented in the code.

In particular for the  `ExtendedKeyUsage`: 
https://github.com/sylabs/singularity/blob/49bb33b0f81c84517d679e18e8324581fa93dd7d/internal/app/singularity/verify_ocsp.go#L112-L115


Is that ok, or do you mean something different?

Reply:
Cool, yes, I think we're good on both those counts now!